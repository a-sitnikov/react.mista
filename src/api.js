//@flow
import fetchJsonp from 'fetch-jsonp'
import * as utils from './utils'

export const domain = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList    = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo     = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin         = localStorage.getItem('urlLogin') || 'ajax_login.php';
export const urlLogout        = localStorage.getItem('urlLogout') || 'users.php'; // users.php?&action=exit
export const urlCookies       = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections      = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage    = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id';
export const urlNewTopic      = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark   = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch        = localStorage.getItem('urlSearch') || 'ajax_find.php';

// Topics list
export type RequestTopicsList = {
    topics?:   number, // число тем, которые нужно вернуть (по умолчанию - 20);
    utime?:    number, // время (в формате unixtime) после которого надо вернуть ветки;
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
    forum?:    string, // раздел форума (1c|it|life);
    section_short_name?: string,
    user_id?:  string,
    mytopics?: string
}

export type ResponseTopicsListItem = {
    id: number,
    forum: string,
    sect1: string,
    sect2: string,
    v8: string,
    closed: number,
    down: number,
    text: string,
    utime: number,
    created: number,
    user: string,
    user0: string,
    is_voting: number,
    answ: number
}

export type ResponseTopicsList = Array<ResponseTopicsListItem>;

export const getTopicsList = async (params: RequestTopicsList): Promise<ResponseTopicsList> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicsList, params);
    return json;
} 


// Topics info
export type RequestInfo = {
    id: string // идентификатор (topic_id) темы
}

export type ResponseVoteItem = {
    select: string,
    result: number
}

export type ResponseInfo = {
    id: string,
    text: string,
    forum?: string,
    section?: string,
    created?: string,
    user_id?: string,
    user_name?: string,
    answers_count: string,
    down?: number,
    closed?: number,
    deleted?: number,
    is_voting?: number,
    voting?: Array<ResponseVoteItem>
}

export const defaultInfo = {
    id: "",
    text: "",
    answers_count: ""
}

export const getTopicInfo = async (params: RequestInfo): Promise<ResponseInfo> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicInfo, params);
    return json;
} 


// Topic messages
export type RequestMessages = {
    id:    number | string, // идентификатор (topic_id) темы
    from?: number, // с какого сообщения. если не указан, то с первого
    to?:   number  // до какого сообщения. если не указан, то from+10
}

export type ResponseMessage = {
    id:    string,
    n:     string,
    user:  string,
    userId: string,
    text:  string,
    utime: string,
    vote:  number
}

export type ResponseMessages = Array<ResponseMessage>;

export const getTopicMessages = async (params: RequestMessages): Promise<ResponseMessages> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicMessages, params);
    return json;
} 

export const getMessage = async (id: number | string, n: number): Promise<?ResponseMessage> =>  {
    const json = await getTopicMessages({
        id, 
        from: n, 
        to: n+1}
        );

    let data;
    if (json && json.length > 0)
        data = json.find(val => val.n === String(n));

    return data;
} 

export const getTopicMessagesCount = async (id: number | string): number => {
    const info = await getTopicInfo({id: String(id)});
    return +info.answers_count;
}

// Login
export type RequestLogin = {
    username:  string, // логин (ник) пользователя
    password:  string, // пароль пользователя
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
}

export type ResponseLogin = {
    userid: string,
    username: string,
    hashkey: string,
    error: string
}

export const getLogin = async (params: RequestLogin): Promise<ResponseLogin> =>  {
    const json = await fetchJsonpAndGetJson(urlLogin, params);
    return json;
} 

export const getLogout = async(): Promise<any> => {
    const params = { action: "exit" };
    await fetchJsonpAndGetJson(urlLogout, params);
}


// Cookies
export type ResponseCookies = {
    cookie: {
        entr_id: string,
        entr_key: string,
        entr_hash: string
    },
    session: {
        user_id: string,
        user_name: string,
        user_hash: string,
        is_moderator: number,
        light_moderator: number,
        section: string,
        show_moderator_tools: number,
        last_error?: string   
    }
}

export const getCookies = async (): Promise<ResponseCookies> =>  {
    const json = await fetchJsonpAndGetJson(urlCookies);
    return json;
} 


// Sections
export type ResponseSection = {
    id: number,
    forum: string,
    shortn: string,
    fulln: string,
    id: number
}

export type ResponseSections = Array<ResponseSection>;

export const getSections = async (): Promise<ResponseSections> =>  {
    const json = await fetchJsonpAndGetJson(urlSections);
    return json;
} 


// New message
export type RequestNewMessage = {
    message_text: string,
    action:       "new", 
    topic_id:     string,
    rnd:          number,
    voting_select?: number,
    as_admin?: boolean    
}

export const postNewMessage = async (params: RequestNewMessage): Promise<any> =>  {

    let url = urlNewMessage.replace(':id', params.topic_id);

    await fetch(`${domain}/${url}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
} 


// New topic
export type RequestNewTopic = {
    action:         "new", 
    rnd:            number,
    topic_text:     string,
    message_text:   string,
    target_section: string, 
    target_forum:   string,
    voting:         number,
    select1?:       string,    
    select2?:       string,    
    select3?:       string,    
    select4?:       string,    
    select5?:       string,    
    select6?:       string,    
    select7?:       string,    
    select8?:       string,    
    select9?:       string,    
    select10?:      string
}

export const postNewTopic = async (params: RequestNewTopic): Promise<any> =>  {

    await fetch(`${domain}/${urlNewTopic}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
        }    
    });
} 


//Bookmark
export type RequestBookmark = {
    id: string
}

export const postBookmark = async (params: RequestBookmark) => {

    await fetch(`${domain}/${urlAddBookmark}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}


//Search
export type RequestSearch = {
    keywords: string
}

export type ResponseSearch = {

}

export const postSearch = async (params: RequestSearch) => {

    await fetch(`${domain}/${urlSearch}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}

export const paramsToString = (paramsPrefix: string, params: ?{}): string => {

    if (!params)
        return '';

    let arr = [];
    for (let key in params) {
        if (params[key] !== undefined)
            arr.push(key + '=' + params[key]);
    }

    if (arr.length > 0)
        return paramsPrefix + arr.join('&');
    else    
        return '';
}

export const fetchJsonpAndGetJson = async (url: string, params: any): Promise<any> => {

    let fullUrl = `${domain}/${url}${paramsToString('?', params)}`; 
    const response =  await fetchJsonp(fullUrl);
    let responseJson = await response.json();
    let json;
    if (typeof(responseJson) === 'string') {
        
        try {
             json = JSON.parse(responseJson)    
        } catch(e) {
            json = utils.parseJSON(responseJson);
        }
    } else {
        json = responseJson;
    }
    return json;
}
