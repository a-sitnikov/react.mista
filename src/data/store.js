import { configureStore  } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import logger from 'redux-logger'

import topicsList from './topicslist/reducer'
import topic from './topic/reducer'
import sections from './sections/reducer'
import options from './options/reducer'
import login from './login/reducer'
import tooltips from './tooltips/reducer'

import newTopic from 'src/reducers/new_topic'
import newMessage from 'src/reducers/new_message'
import topicPreview from './topic_preview/reducer'

const reducer = {
  topicsList,
  topic,
  sections,
  options,
  login,
  tooltips,
  newTopic,
  newMessage,
  topicPreview  
}

const store = configureStore({
  reducer,
  //middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production'
})

export const useAppDispatch = () => useDispatch()

export default store