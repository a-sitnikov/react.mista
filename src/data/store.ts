import { configureStore  } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import logger from 'redux-logger' // eslint-disable-line @typescript-eslint/no-unused-vars

import topicsList from './topicslist/reducer'
import topic from './topic/reducer'
import sections from './sections/reducer'
import options from './options/reducer'
import login from './login/reducer'
import tooltips from './tooltips/reducer'
import newTopic from './newtopic/reducer'
import newMessage from './newmessage/reducer'

const reducer = {
  topicsList: topicsList.reducer,
  topic,
  sections: sections.reducer,
  options,
  login,
  tooltips: tooltips.reducer,
  newTopic,
  newMessage
}

const store = configureStore({
  reducer,
  //middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  // middleware: (getDefaultMiddleware) => getDefaultMiddleware({
  //   immutableCheck: { warnAfter: 128 },
  //   serializableCheck: { warnAfter: 128 },
  // })
  // .concat(logger)
  devTools: process.env.NODE_ENV !== 'production'
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export default store