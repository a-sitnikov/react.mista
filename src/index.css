@import "tailwindcss";
@config '../tailwind.config.ts';

@custom-variant dark (&:where([data-theme="dark"], [data-theme="dark"] *));

@theme {
  --color-bgHeader: #eee;
  --color-borderColor: #ccc;
  --color-bgColor: white;
  --color-bgHover: #eee;
  --color-link: #00c;
}

@utility border-outer {
  @apply border-1 border-(--color-borderColor);
}

:root {
  --border-color: #ccc;
  --inner-border-color: #eee;
  --header-background: #eee;
  --background: white;
  --background-hover: #f5f9fc;
  --font-color: black;
  --link-color: #00c;
  --link-hover-color: #d00;
  --mytopics-color: #052;
  --pinned-color: purple;
  --button-hover-bg-color: var(--bs-light);
  --button-disabled-color: #888;
  --button-hover-border-color: #eee #888 #888 #eee;
}

[data-theme="dark"] {
  --color-bgHeader: black;
  --color-borderColor: rgb(100, 100, 100);
  --color-bgColor: black;
  --color-bgHover: rgb(50, 50, 50);

  --border-color: rgb(100, 100, 100);
  --inner-border-color: rgb(100, 100, 100);
  --header-background: black;
  --background: black;
  --background-hover: rgb(25, 25, 25);
  --font-color: white;
  --link-color: #00a2fd;
  --link-hover-color: #d00;
  --mytopics-color: #052;
  --pinned-color: rgb(207, 41, 207);
  --button-hover-bg-color: rgb(15, 15, 15);
  --button-disabled-color: rgb(100, 100, 100);
  --button-hover-border-color: #eee #888 #888 #eee;
}

html {
  min-height: 100%;
}

body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
  font-size: initial !important;
  /*color: black !important*/
}

body[data-theme="yellow"] {
  background-image: url("/img/fon_forum.gif");
}

body[data-theme="lightgray"] {
  background-image: url("/img/white_carbon.png");
}

[data-theme="dark"] {
  background-color: black;
  color: var(--font-color);
}

#root {
  min-height: 100vh;
  position: relative;
  display: flex;
  flex-direction: column;
}

.flex-row {
  width: 100%;
  display: flex;
}

.flex-column {
  width: 100%;
  display: flex;
  flex-direction: column;
}

.separator {
  margin-left: 5px;
  margin-right: 5px;
  user-select: none;
}

.div-va-middle {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
}

.button {
  color: var(--font-color);
  background: var(--background);
  border: 1px solid var(--border-color);
  border-radius: 3px;
}

.button:hover:not([disabled=""]) {
  color: var(--font-color);
  background: var(--button-hover-bg-color);
  border-color: var(--button-hover-border-color);
  /*background: var(--background);*/
}

.button:disabled {
  color: var(--button-disabled-color);
}

.dropdown-toggle {
  background-color: var(--background);
  border-color: var(--border-color);
  color: var(--font-color);
}

[data-theme="dark"] .dropdown-menu {
  background-color: var(--background);
  border: 1px solid var(--border-color);
}

[data-theme="dark"] .dropdown-item {
  color: var(--font-color);
}

[data-theme="dark"] .bg-dark {
  background-color: var(--background) !important;
}

.ah {
  text-decoration: none;
  font-size: 80%;
}

.agh {
  font-size: 80%;
  text-decoration: none;
  color: gray;
}

.mr5 {
  margin-right: 5px;
}

.bold {
  font-weight: bold;
}

.cc {
  text-align: center;
  vertical-align: top;
  font-size: 80%;
}

.ct {
  text-align: left;
  vertical-align: top;
}

.cl {
  text-align: left;
  vertical-align: top;
  font-size: 80%;
  word-wrap: break-word;
}

a {
  color: var(--link-color);
  text-decoration: none !important;
}

a:hover {
  color: var(--link-hover-color);
  text-decoration: unset;
}

.your-name {
  color: var(--link-color);
}

.link {
  cursor: pointer;
  color: var(--link-color);
}

.mytopics {
  color: var(--mytopics-color);
}

.va-middle {
  vertical-align: middle !important;
}

.ta-right {
  text-align: right !important;
}

.va-top {
  vertical-align: top !important;
}

.ta-left {
  text-align: left !important;
}

.ta-center {
  text-align: center !important;
}

.bold120 {
  color: #309;
  font-weight: bold;
  font-size: 120%;
}

.form-control {
  /*transition: none !important;*/
  box-shadow: none !important;
  border-color: var(--border-color);
}

.form-control:focus {
  box-shadow: none !important;
}

.container {
  margin-top: 45px;
  flex-grow: 1;
}

.input {
  color: var(--font-color);
  border-color: var(--border-color);
}

[data-theme="dark"] .input {
  background: black;
  color: inherit;
}

@media (max-width: 768px) {
  .container-fluid {
    padding-right: 0px;
    padding-left: 0px;
  }

  .container-fluid > .navbar-collapse {
    margin-left: 3px;
    margin-right: 3px;
  }

  .container {
    margin-top: 60px;
    padding-left: 3px;
    padding-right: 3px;
  }
}

@media (min-width: 1400px) {
  .container {
    max-width: 1156px;
  }
}
