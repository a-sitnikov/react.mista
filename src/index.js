//@flow
import React from 'react';
import ReactDOM from 'react-dom';
import registerServiceWorker from './registerServiceWorker';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux'

import thunk from 'redux-thunk'

// eslint-disable-next-line 
import { createLogger } from 'redux-logger'

import reducer from './reducers'
import './index.css';
import App from './App';

const middleware = [thunk];

if (process.env.NODE_ENV !== 'production') {
    //middleware.push(createLogger())
}

const store = createStore(
    reducer,
    applyMiddleware(...middleware)
)

const Root = ({ store }) => (
    <Provider store={store}>
        <App/>
    </Provider>
)

// for github pages
window.hash = "#";

ReactDOM.render(<Root store={store} />,
    document.getElementById('root'));

registerServiceWorker();
