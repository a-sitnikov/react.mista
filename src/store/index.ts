export { reducer, store, persistor } from "./root";
export { useAppSelector, useAppDispatch, useActionCreators } from "./hooks";
export type { RootState, AppDispatch } from "./types";

export * from "./slices";
