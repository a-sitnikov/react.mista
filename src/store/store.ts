import { configureStore  } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import logger from 'redux-logger' // eslint-disable-line @typescript-eslint/no-unused-vars

import { topicsList } from './topics_list'
import { topic } from './topic'
import { sections } from './sections'
import { options } from './options'
import { login } from './login'
import { tooltips } from './tooltips'
import { newTopic } from './new_topic'
import newMessage from '../data/newmessage/reducer'

const reducer = {
  topicsList,
  topic,
  sections,
  options,
  login,
  tooltips,
  newTopic,
  newMessage
}

const store = configureStore({
  reducer,
  //middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production'
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export default store