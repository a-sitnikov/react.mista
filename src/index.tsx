import { createRoot } from 'react-dom/client';
import { Provider } from 'react-redux'

import './index.css';
import App from './App';
import { store } from 'src/store'

const Root = ({ store }): React.ReactElement => (
  <Provider store={store}>
    <App />
  </Provider>
)

const container = document.getElementById('root');
const root = createRoot(container!);
root.render(<Root store={store} />);