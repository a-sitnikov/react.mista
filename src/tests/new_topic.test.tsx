import { fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import { setupStore, renderWithProviders } from './test-utils';
import NewTopic from 'src/pages/topics_list/new_topic'
import { NewTopicState, SectionsState } from 'src/store';
import { mock_sections, mock_sections_tree } from './mock_data';

describe('NewTopic', () => {

  const sections: SectionsState = {
    status: "init",
    items: mock_sections,
    tree: mock_sections_tree
  }

  const newTopic: NewTopicState = {
    status: "init",
    section: null,
    text: '',
    subject: '',
    forum: '1C',
    isVoting: false
  }

  let initialState = {
    sections,
    newTopic
  }

  beforeAll(() => {
  })

  it('render', () => {

    const store = setupStore(initialState);
    renderWithProviders(
      <NewTopic />,
      { store }
    );

    expect(screen.getByRole('textbox', { name: "Тема" })).toBeInTheDocument();
    expect(screen.getByRole('textbox', { name: "Сообщение" })).toBeInTheDocument();

  });

  it('render 10 voting options', () => {

    const store = setupStore(initialState);
    renderWithProviders(
      <NewTopic />,
      { store }
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
    
    fireEvent.click(checkbox);

    expect(checkbox).toBeChecked();

    const inputs = screen.getAllByRole('textbox', { name: /Вариант/});
    expect(inputs.length).toBe(10);
    expect(inputs[0]).toBeInTheDocument();

  });

})    