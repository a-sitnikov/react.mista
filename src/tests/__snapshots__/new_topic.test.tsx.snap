// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTopic render 1`] = `
<div>
  <form
    class="new-topic-container"
  >
    <div
      class="new-topic-text"
      id="newtopic_form"
    >
      <div>
        <b>
          Новая тема:
        </b>
      </div>
      <div
        class="flex-row"
        style="margin-bottom: 3px;"
      >
        <select
          class="input form-select form-select-sm"
          disabled=""
          style="flex: 0 1 90px;"
        >
          <option
            value="1c"
          >
            1C
          </option>
          <option
            value="it"
          >
            IT
          </option>
        </select>
        <select
          aria-label="Секция"
          class="input form-select form-select-sm"
          id="target_section"
          style="flex: 1 1 auto;"
        >
          <option
            value=""
          >
            Секция
          </option>
          <optgroup
            label="1C"
          >
            <option
              value="v7"
            >
              1С:Предприятие 7.7 и ранее
            </option>
            <option
              value="v8"
            >
              1С:Предприятие 8 общая
            </option>
          </optgroup>
          <optgroup
            label="IT"
          >
            <option
              value="admin"
            >
              Администрирование
            </option>
            <option
              value="mobile"
            >
              Мобильный мир
            </option>
          </optgroup>
        </select>
      </div>
      <input
        aria-label="ttt"
        class="input form-control form-control-sm"
        maxlength="90"
        placeholder="Тема"
        style="margin-bottom: 3px;"
        type="text"
        value=""
      />
      <div>
        <textarea
          aria-label="Сообщение"
          class="text-editor input form-control"
          cols="70"
          data-lpignore="true"
          placeholder="Сообщение"
          rows="3"
        />
        <div
          class="flex-row"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="button btn btn-light btn-sm"
              style="margin-right: 5px;"
              type="button"
            >
              Код 1С
            </button>
            <button
              class="button btn btn-light btn-sm"
              type="submit"
            >
              Отправить
            </button>
          </div>
          <div
            class="form-check"
            style="margin: auto 0px auto auto;"
          >
            <input
              class="form-check-input"
              id="show_voting"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="show_voting"
              title=""
            >
              Голосование
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="new-topic-voting"
    />
  </form>
</div>
`;
